light-clients/icsxx-cf-solana/src/header.rs

```diff
@@ -1,2 +1,397 @@
-super::wrap!(cf_guest_upstream::Header<PK> as Header);
-super::wrap!(impl<PK> proto for Header);
+use crate::{
+	proto,
+	solana::{
+		blockstore::{get_completed_data_ranges, get_slot_entries_in_block},
+		shred::{shred_code::ShredCode, shred_data::ShredData, *},
+	},
+};
+use alloc::vec::Vec;
+use guestchain::PubKey;
+use proto_utils::BadMessage;
+use solana_sdk::{clock::Slot, hash::Hash, signature::Signature};
+use std::{collections::BTreeSet, convert::From, marker::PhantomData};
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct Header<PK> {
+	pub shreds: Vec<Shred>,
+	_phantom: PhantomData<PK>,
+}
+
+impl<PK> Header<PK> {
+	pub(crate) fn slot(&self) -> Slot {
+		assert!(!self.shreds.is_empty(), "Header must contain at least one shred");
+		self.shreds[0].slot()
+	}
+
+	pub fn hash(&self) -> Hash {
+		let mut shreds = self.shreds.iter().collect::<Vec<_>>();
```

Comment:
Unless I’m missing something, `shreds` is only used to make
`data_shreds`.  Based on that we probably should do filtering here so
that we don’t allocate unnecessary empty vector:

```suggestion
		let mut data_shreds = self.shreds
			.iter()
			.filter(|shred| shred.is_data())
			.collect::<Vec<_>>();
		data_shreds.sort_by_key(|s| s.index());
		data_shreds.dedup_by_key(|s| s.index());
```

Reply:
Removed the code because the checks are done in `PreCheckedShreds` now